### PR TITLE
New URL parser

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
When the issue 4717 was created(https://github.com/netdata/netdata/issues/4717 ) it was done the suggestion to rewrite the url parser, this fix has this in mind. I am working not only to speed up the whole parser, but to eliminate unnecessary copy of buffers.

##### Component Name
This fix will affect for while the Web server, but I saw that the stream also parses URLs, so it is necessary to extend for it too.
##### Additional Information
To test the newest URL, I have done the following steps:

1 - I am using a new script that I am bringing together with this fix, it was stored inside tests/urls it its name is requests.sh, please, pay attention for the fact that it also tests manage/health, so to execute the last test it will be necessary to have a TOKEN. All the tests with this scripts will be stored in two directories, 
2 - Normal access trough browsers to confirm that everything is there.

### CODE REVIEW

File: SECURITY.md [hunk 1]
```diff
@@ -1,43 +0,0 @@
-# Security Policy
```
Comment (cakrit) [inline SECURITY.md]: This file shouldn't be here. There's a problem with your branch.

File: libnetdata/inlined.h [hunk 2]
```diff
@@ -31,6 +31,19 @@ static inline uint32_t simple_hash(const char *name) {
...
 
+static inline uint32_t simple_nhash(const char *name,size_t len) {
+    unsigned char *s = (unsigned char *) name;
+    size_t i;
+    uint32_t hval = 0x811c9dc5;
+    i = 0;
+    do {
+        hval *= 16777619;
+        hval ^= (uint32_t) *s++;
+    }while (++i < len);
```
Comment (cakrit) [inline libnetdata/inlined.h]: Space after `}'

File: libnetdata/inlined.h [hunk 3]
```diff
@@ -42,6 +55,21 @@ static inline uint32_t simple_uhash(const char *name) {
...
+    size_t i;
+    uint32_t hval = 0x811c9dc5, c;
+
+    i = 0;
+    do {
+        c = *s++;
+        if (unlikely(c >= 'A' && c <= 'Z')) c += 'a' - 'A';
+        hval *= 16777619;
+        hval ^= c;
+    }while ( ++i < len);
```
Comment (cakrit) [inline libnetdata/inlined.h]: Space after '}'

File: libnetdata/socket/security.h [hunk 4]
```diff
@@ -18,7 +18,7 @@
 #define NETDATA_SSL_VALID_CERTIFICATE 128  //Accepts invalid certificate
 struct netdata_ssl{
     SSL *conn; //SSL connection
-    int flags;
+    uint32_t flags;
```
Comment (cakrit) [inline libnetdata/socket/security.h]: This change doesn't belong here.

Comment (thiagoftsm) [reply libnetdata/socket/security.h]: I will create a new PR for this, while I was compiling with Clion it showed some warnings due clang compiler and I changed here.

File: web/api/badges/web_buffer_svg.c [hunk 5]
```diff
@@ -912,46 +913,54 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
...
-        else if(!strcmp(name, "refresh")) refresh_str = value;
-        else if(!strcmp(name, "precision")) precision_str = value;
-        else if(!strcmp(name, "scale")) scale_str = value;
-        else if(!strcmp(name, "alarm")) alarm = value;
+    uint32_t i = 0;
+    uint32_t end = w->total_params;
+    char save[WEB_FIELDS_MAX];
+    char *value;
+    size_t lvalue;
+    if(end){
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space before '{'

File: web/api/badges/web_buffer_svg.c [hunk 6]
```diff
@@ -912,46 +913,54 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
...
+        do{
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space before '{'

File: web/api/badges/web_buffer_svg.c [hunk 7]
```diff
@@ -912,46 +913,54 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
...
+            else if(!strncmp(name, "label_color",lname)) label_color = value;
+            else if(!strncmp(name, "value_color",lname)) value_color = value;
+            else if(!strncmp(name, "multiply",lname)) multiply_str = value;
+            else if(!strncmp(name, "divide",lname)) divide_str = value;
+            else if(!strncmp(name, "refresh",lname)) refresh_str = value;
+            else if(!strncmp(name, "precision",lname)) precision_str = value;
+            else if(!strncmp(name, "scale",lname)) scale_str = value;
+            else if(!strncmp(name, "alarm",lname)) alarm = value;
+
+        }while (++i < end );
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space after '}'

File: web/api/badges/web_buffer_svg.c [hunk 8]
```diff
@@ -1137,6 +1146,14 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
     }
 
     cleanup:
+    if(end){
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space before '{'

File: web/api/badges/web_buffer_svg.c [hunk 9]
```diff
@@ -1137,6 +1146,14 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
...
+        i = 0;
+        do{
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space before '{'

File: web/api/badges/web_buffer_svg.c [hunk 10]
```diff
@@ -1137,6 +1146,14 @@ int web_client_api_request_v1_badge(RRDHOST *host, struct web_client *w, char *u
...
+            value = w->param_values[i].body;
+            lvalue = w->param_values[i].length;
+            value[lvalue] = save[i];
+        }while(++i < end);
```
Comment (cakrit) [inline web/api/badges/web_buffer_svg.c]: space after }

File: web/api/exporters/allmetrics.c [hunk 11]
```diff
@@ -24,48 +24,50 @@ inline int web_client_api_request_v1_allmetrics(RRDHOST *host, struct web_client
     PROMETHEUS_OUTPUT_OPTIONS prometheus_output_options = PROMETHEUS_OUTPUT_TIMESTAMPS | ((global_backend_options & BACKEND_OPTION_SEND_NAMES)?PROMETHEUS_OUTPUT_NAMES:0);
     const char *prometheus_prefix = global_backend_prefix;
 
-    while(url) {
-        char *value = mystrsep(&url, "&");
-        if (!value || !*value) continue;
+    uint32_t end = w->total_params;
+    if (end){
```
Comment (cakrit) [inline web/api/exporters/allmetrics.c]: spaces here too

File: SECURITY.md (hunk 1 above)
Comment (thiagoftsm) [reply SECURITY.md]: Thanks, I am fixing now!

File: libnetdata/inlined.h (hunk 2 above)
Comment (thiagoftsm) [reply libnetdata/inlined.h]: Next pull will fix this!

File: libnetdata/inlined.h (hunk 3 above)
Comment (thiagoftsm) [reply libnetdata/inlined.h]: Next pull will fix this!

File: libnetdata/socket/security.h (hunk 4 above)
Comment (thiagoftsm) [reply libnetdata/socket/security.h]: Next pull will fix this!

File: web/server/web_client_cache.c [hunk 12]
```diff
@@ -188,15 +188,15 @@ struct web_client *web_client_get_from_cache_or_allocate() {
 #ifdef ENABLE_HTTPS
```
Comment (cakrit) [inline web/server/web_client_cache.c]: Changes here don't belong in this PR

File: web/api/badges/web_buffer_svg.c (hunk 5 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: Next pull will fix this! I changed it and the do near to it.

File: web/api/badges/web_buffer_svg.c (hunk 6 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: Next pull will fix this!

File: web/api/badges/web_buffer_svg.c (hunk 7 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: Next pull will fix this!

File: web/api/badges/web_buffer_svg.c (hunk 8 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: I am reviewing all the files changed to put the space, I already got the rule!

File: web/api/badges/web_buffer_svg.c (hunk 9 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: Next pull will fix this!

File: web/api/badges/web_buffer_svg.c (hunk 10 above)
Comment (thiagoftsm) [reply web/api/badges/web_buffer_svg.c]: Next pull will fix this!

File: web/api/exporters/allmetrics.c [hunk 13]
```diff
@@ -24,48 +24,50 @@ inline int web_client_api_request_v1_allmetrics(RRDHOST *host, struct web_client
     PROMETHEUS_OUTPUT_OPTIONS prometheus_output_options = PROMETHEUS_OUTPUT_TIMESTAMPS | ((global_backend_options & BACKEND_OPTION_SEND_NAMES)?PROMETHEUS_OUTPUT_NAMES:0);
```
Comment (cakrit) [inline web/api/exporters/allmetrics.c]:
```
/home/christopher/CLionProjects/netdata/web/api/exporters/allmetrics.c: In function ‘web_client_api_request_v1_allmetrics’:
/home/christopher/CLionProjects/netdata/web/api/exporters/allmetrics.c:20:92: warning: unused parameter ‘url’ [-Wunused-parameter]
 inline int web_client_api_request_v1_allmetrics(RRDHOST *host, struct web_client *w, char *url) {
```

File: web/server/web_client_cache.c (hunk 12 above)
Comment (thiagoftsm) [reply web/server/web_client_cache.c]: I compared the master with this, and it looks like the change was only space, because this line is present in the master.

File: web/api/exporters/allmetrics.c (hunk 11 above)
Comment (thiagoftsm) [reply web/api/exporters/allmetrics.c]: Next pull will fix this!

File: web/api/exporters/allmetrics.c (hunk 13 above)
Comment (thiagoftsm) [reply web/api/exporters/allmetrics.c]: Next pull will fix this!

File: web/server/web_client_cache.c (hunk 12 above)
Comment (thiagoftsm) [reply web/server/web_client_cache.c]: I remembered the changes that I made, I executed a new pull for it!

File: SECURITY.md (hunk 1 above)
Comment (thiagoftsm) [reply SECURITY.md]: I deleted the file, but I had to return it. This file was created in a "git pull" that  I executed and without it the github is giving errors.

Comment (cakrit) [reply SECURITY.md]: This deletes the file though! Try a rebase.

File: web/api/exporters/allmetrics.c [hunk 14]
```diff
@@ -18,54 +18,57 @@ struct prometheus_output_options {
...
 
-    while(url) {
-        char *value = mystrsep(&url, "&");
-        if (!value || !*value) continue;
+    uint32_t end = w->total_params;
+    if (end) {
+        uint32_t i = 0;
+        do {
+            char *name = w->param_name[i].body;
+            size_t lname = w->param_name[i].length;
```
Comment (cakrit) [inline web/api/exporters/allmetrics.c]: typo -> length

File: web/api/web_api_v1.c [hunk 15]
```diff
@@ -195,14 +195,19 @@ inline uint32_t web_client_api_request_v1_data_google_format(char *name) {
...
+    uint32_t end = w->total_params;
+    if(end) {
+        uint32_t  i = 0;
+        do {
+            char *value = w->param_values[i].body;
+            size_t length = w->param_values[i].length;
 
-        if(!strcmp(value, "all")) all = 1;
-        else if(!strcmp(value, "active")) all = 0;
+            if(!strncmp(value, "all",length)) all = 1;
```
Comment (cakrit) [inline web/api/web_api_v1.c]: Something doesn't work with this parameter. When you send https://localhost:19999/api/v1/alarms?all, you're supposed to get all the configured alarms. Instead, we get all the raised alarms, i.e. exactly the same as if we don't pass "all" at all.

File: web/api/health/health_cmdapi.c [hunk 16]
```diff
@@ -73,75 +70,85 @@ int web_client_api_request_v1_mgmt_health(RRDHOST *host, struct web_client *w, c
...
-                        if (
-                                (hash == hash_alarm && !strcasecmp(key, HEALTH_ALARM_KEY)) ||
-                                (hash == hash_template && !strcasecmp(key, HEALTH_TEMPLATE_KEY)) ||
-                                (hash == hash_chart && !strcasecmp(key, HEALTH_CHART_KEY)) ||
-                                (hash == hash_context && !strcasecmp(key, HEALTH_CONTEXT_KEY)) ||
-                                (hash == hash_host && !strcasecmp(key, HEALTH_HOST_KEY)) ||
-                                (hash == hash_families && !strcasecmp(key, HEALTH_FAMILIES_KEY))
-                                ) {
-                            silencer = create_silencer();
+            uint32_t end = w->total_params;
```
Comment (cakrit) [inline web/api/health/health_cmdapi.c]: Something doesn't work here.

```
Cmd 'cmd=SILENCE&alarm=*example_alarm1 *load_trigger'
  Expecting 'Alarm notifications silenced for alarms matching the selectors
Alarm selector added' : ERROR: Response 'All alarm notifications are silenced
Invalid key. Ignoring it.'
```
I'll do a PR to fix some things with `health-cmdapi-test.sh` so you can run it too, to repeat the tests.

File: web/api/web_api_v1.c (hunk 15 above)
Comment (thiagoftsm) [reply web/api/web_api_v1.c]: I fixed the error, there was a bug in parser this kind of query strings.

File: web/api/health/health_cmdapi.c (hunk 16 above)
Comment (thiagoftsm) [reply web/api/health/health_cmdapi.c]: I tested few commands of this file and I saw that it could work, I am waiting you to repeat here.

File: SECURITY.md (hunk 1 above)
Comment (thiagoftsm) [reply SECURITY.md]: It is fixed in the last pull! Thanks!

File: web/api/web_api_v1.c (hunk 15 above)
Comment (cakrit) [reply web/api/web_api_v1.c]: Not fixed completely. Now the following doesn't work properly:
http://localhost:19999/api/v1/alarms?all&132111

File: web/api/health/health_cmdapi.c (hunk 16 above)
Comment (cakrit) [reply web/api/health/health_cmdapi.c]: Cmd 'cmd=SILENCE&alarm=*10min_cpu_usage *load_trigger'
  Expecting 'Alarm notifications silenced for alarms matching the selectors
Alarm selector added' : ERROR: Response 'All alarm notifications are silenced
Invalid key. Ignoring it.'

Comment (thiagoftsm) [reply web/api/health/health_cmdapi.c]: There was a previous error that avoided the system to execute correctly the "if/else if" algorithm, but now I could check that we are going in all the options as it expected, but I got 15 errors during script execution, please, can you check the arguments that are given for Netdata?

Comment (thiagoftsm) [reply web/api/health/health_cmdapi.c]: Problem completely fixed! I executed both scripts and there was not any error.

File: libnetdata/url/url.c [hunk 17]
```diff
@@ -79,3 +79,88 @@ char *url_decode_r(char *to, char *url, size_t size) {
 
     return to;
 }
+
+inline HTTP_VALIDATION url_is_request_complete(char *begin,char *end,size_t length){
```
Comment (ilyam8) [inline libnetdata/url/url.c]: please add spaces between func arguments

true for all funcs

File: libnetdata/url/url.c [hunk 18]
```diff
@@ -80,18 +80,18 @@ char *url_decode_r(char *to, char *url, size_t size) {
...
 
-inline HTTP_VALIDATION url_is_request_complete(char *begin,char *end,size_t length){
-    if ( begin == end){
+inline HTTP_VALIDATION url_is_request_complete(char *begin,char *end,size_t length) {
+    if ( begin == end) {
         return HTTP_VALIDATION_INCOMPLETE;
     }
 
-    if ( length > 3  ){
+    if ( length > 3  ) {
```
Comment (ilyam8) [inline libnetdata/url/url.c]: you are not consistent with spaces in condition, i think there shouldnt be any trailing and leading spaces according the guide we accepted

Comment (thiagoftsm) [reply libnetdata/url/url.c]: The newest pull fixed this!

**Comment by thiagoftsm** [comment]
I am already working to fix it...
